### PR TITLE
[LoRA] Parallelize rank dimension in embedding_lora_a kernel for up to 1.36x speedup

### PR DESCRIPTION
**Pull request by flutist**
## Motivation

The `embedding_lora_a` Triton kernel processes the rank dimension in a **serial loop** inside each program:

```python
# Before: serial loop over rank blocks
for block_id in range(num_blocks):
    rank_offset = tl.arange(0, BLOCK_RANK) + block_id * BLOCK_RANK
    ...
```

When `rank > BLOCK_RANK (128)`, multiple iterations execute sequentially within a single GPU thread block, leaving GPU parallelism on the table. This is the same pattern that `chunked_sgmv_shrink.py` already parallelizes via its grid N‑dimension.

### Kernel optimization (`python/sglang/srt/lora/triton_ops/embedding_lora_a.py`)

- **Replaced the serial `for` loop** with a third grid dimension that parallelizes rank blocks across GPU thread blocks
- Grid changed from `(max_len, bs)` → `(max_len, bs, cdiv(rank, BLOCK_RANK))`
- Each program now handles exactly one rank block via `rank_block_id = tl.program_id(axis=2)`
- Added early‑exit guard: `if rank_block_id * BLOCK_RANK >= rank_val: return`
- **No interface change**: `embedding_lora_a_fwd()` signature is unchanged; callers (`triton_backend.py`) require zero modifications

### New correctness test (`test/registered/lora/test_embedding_lora_a.py`)

- pytest style following `test_fused_moe_lora_kernel.py`
- 28 test cases covering:
  - Basic batch sizes
  - Decode mode (seq_len=1)
  - Single / multiple rank blocks
  - Non‑aligned ranks (100/200/300)
  - Mixed ranks across adapters
  - Rank=0 edge cases
  - Unequal segment lengths
- Registered to `stage-b-test-1-gpu-small` CI suite

### New benchmark (`benchmark/lora/bench_embedding_lora_a.py`)

- `triton.testing.perf_report` style following `jit_kernel/benchmark/bench_activation.py`
- Sweeps batch_size × seq_len × rank combinations

## Accuracy Tests

**28/28 passed** ✅

```bash
$ python -m pytest test/registered/lora/test_embedding_lora_a.py -v
========================= 28 passed, 17 warnings in 7.61s =========================
```

## Speed Tests and Profiling

| batch_size | seq_len | rank | Before (µs) | After (µs) | Speedup |
|------------|---------|------|-------------|------------|---------|
| 8          | 512     | 256  | 122.90      | 100.70     | **1.22x** |
| 32         | 128     | 256  | 151.31      | 131.92     | **1.15x** |
| 32         | 512     | 256  | 295.10      | 216.90     | **1.36x** |

- **rank ≤ 128**: Performance unchanged (single rank block)
- **rank = 256 + large batch**: Up to **1.36x speedup**
- **No regression** at any configuration

---
### CI States

Latest PR Test (Base): :x: [Run #26808801748](https://github.com/sgl-project/sglang/actions/runs/26808801748)
Latest PR Test (Extra): :x: [Run #26808801541](https://github.com/sgl-project/sglang/actions/runs/26808801541)

### CODE REVIEW

**Comment by flutist** [comment]
@Fridge003 Sorry to bother you, but could you please help me merge this PR file? This solved the problem. If there's anything else I can do, I'll continue. I'm very happy to hear your response.

**Comment by flutist** [comment]
@yushengsu-thu Sorry to bother you, but could you please help me merge this PR file? This solved the problem. If there's anything else I can do, I'll continue. I'm very happy to hear your response.

File: test/registered/lora/test_embedding_lora_a.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+    # --- PyTorch reference (same one used by test_chunked_sgmv_backend) ---
+    reference_output = reference_embedding_lora_a_shrink(
+        input_ids,
+        weights,
+        torch.tensor(lora_assignments, dtype=torch.int32, device="cpu"),
+        torch.tensor(seq_lengths, dtype=torch.int32, device="cpu"),
+        batch_info.lora_ranks.detach().cpu(),
+        vocab_size,
+    )
```
[the remark this answers is not in the document]
Comment (flutist) [reply test/registered/lora/test_embedding_lora_a.py:133]: fix

**Comment by flutist** [comment]
@Fridge003 @yushengsu-thu @Ying1123 @lifuhuang [LoRA] Parallelize rank dimension in embedding_lora_a kernel for up to 1.36x speedup
Sorry to bother you, but could you please help me merge this PR file? This solved the problem. If there's anything else I can do, I'll continue. I'm very happy to hear your response.

**Comment by flutist** [comment]
@Fridge003 @yushengsu-thu @Ying1123 @lifuhuang [LoRA] Parallelize rank dimension in embedding_lora_a kernel for up to 1.36x speedup
Sorry to bother you, but could you please help me merge this PR file? This solved the problem. If there's anything else I can do, I'll continue. I'm very happy to hear your response.
